testutils/clustermanager/gke.go

```diff
@@ -18,54 +18,140 @@ package clustermanager
+type GKESDKClient struct {
+	*container.Service
+}
+
+func (gsc *GKESDKClient) create(project, location string, rb *container.CreateClusterRequest) (*container.Operation, error) {
+	parent := fmt.Sprintf("projects/%s/locations/%s", project, location)
+	return gsc.Projects.Locations.Clusters.Create(parent, rb).Context(context.Background()).Do()
+}
+
+func (gsc *GKESDKClient) get(project, location, cluster string) (*container.Cluster, error) {
+	clusterFullPath := fmt.Sprintf("projects/%s/locations/%s/clusters/%s", project, location, cluster)
+	return gsc.Projects.Locations.Clusters.Get(clusterFullPath).Context(context.Background()).Do()
 }
 
 // Setup sets up a GKECluster client.
 // numNodes: default to 3 if not provided
 // nodeType: default to n1-standard-4 if not provided
 // region: default to regional cluster if not provided, and use default backup regions
 // zone: default is none, must be provided together with region
-func (gs *GKEClient) Setup(numNodes *int, nodeType *string, region *string, zone *string, project *string) (ClusterOperations, error) {
-	gc := &GKECluster{}
-	// check for local run
-	if nil != project { // if project is supplied, use it and create cluster
+func (gs *GKEClient) Setup(numNodes *int64, nodeType *string, region *string, zone *string, project *string) (ClusterOperations, error) {
+	var err error
+	gc := &GKECluster{
+		Request: &GKERequest{
+			NumNodes:      DefaultGKENumNodes,
+			NodeType:      DefaultGKENodeType,
+			Region:        DefaultGKERegion,
+			Zone:          DefaultGKEZone,
+			BackupRegions: DefaultGKEBackupRegions},
+	}
+
+	ctx := context.Background()
+
+	c, err := google.DefaultClient(ctx, container.CloudPlatformScope)
+	if nil != err {
+		return nil, fmt.Errorf("failed create google client: '%v'", err)
+	}
+
+	containerService, err := container.New(c)
+	if nil != err {
+		return nil, fmt.Errorf("failed create google client: '%v'", err)
+	}
+	gc.oprations = &GKESDKClient{containerService}
+
+	if nil != project { // use provided project and create cluster
 		gc.Project = project
 		gc.NeedCleanup = true
 	} else if err := gc.checkEnvironment(); nil != err {
 		return nil, fmt.Errorf("failed checking existing cluster: '%v'", err)
 	} else if nil != gc.Cluster { // return if Cluster was already set by kubeconfig
 		return gc, nil
 	}
-	// check for Prow
-	if common.IsProw() {
-		project, err := boskos.AcquireGKEProject()
-		if nil != err {
-			return nil, fmt.Errorf("failed acquire boskos project: '%v'", err)
+	if nil == gc.Cluster {
+		if common.IsProw() {
+			project, err := boskos.AcquireGKEProject()
+			if nil != err {
+				return nil, fmt.Errorf("failed acquire boskos project: '%v'", err)
+			}
+			gc.Project = &project
+		}
+		if nil != numNodes {
+			gc.Request.NumNodes = *numNodes
+		}
+		if nil != nodeType {
+			gc.Request.NodeType = *nodeType
+		}
+		if nil != region {
+			gc.Request.Region = *region
+			gc.Request.BackupRegions = make([]string, 0)
```

Comment:
Can we be smarter about the regions here? A simple approach would be "use the default plus default backup regions, except one that was specified by the user". So, if  I use us-east1 here, central1 and west1 would become the backup regions.

Reply:
Why? A project with a very specific region is pretty custom, so it would make sense controlling every aspect (and in this case, it just sets E2E_BACKUP_REGION to empty). And actually I'm thinking more of those cases where we want to shift the default region to some other because of stockout/performance/etc, as it happened in the past.

Reply:
I see the point, will update it

Reply:
done